src/Symfony/Component/Lock/Store/ZookeeperStore.php

```diff
@@ -0,0 +1,239 @@
+use Symfony\Component\Lock\StoreInterface;
+use Zookeeper;
+use Throwable;
+
+/**
+ * ZookeeperStore is a StoreInterface implementation using Zookeeper as store engine.
+ *
+ * @author Ganesh Chandrasekaran <gchandrasekaran@wayfair.com>
+ */
+class ZookeeperStore implements StoreInterface
+{
+    /**
+     * @var \Zookeeper
+     */
+    private $zookeeper;
+
+    /**
+     * @var string A prefix for the locking resource to identify the process
+     */
+    private $prefix;
+
+    /**
+     * A base node within which all the locks for PHP application are managed in the Zookeeper environment.
+     */
+    private const LOCK_NODE = '/PHP_LOCK_NODE';
+
+    /**
+     * Locker constructor.
+     *
+     * @param \Zookeeper $zookeeper  Zookeeper Instance
+     * @param string     $lockPrefix Prefix for a locking resource
+     */
+    public function __construct(
+        Zookeeper $zookeeper,
+        string $lockPrefix = ''
+    ) {
+        $this->zookeeper = $zookeeper;
+        $this->prefix = $lockPrefix;
+    }
+
+    /**
+     * Stores the resource if it's not locked by someone else.
+     *
+     * @param \Symfony\Component\Lock\Key $key the Key that represent the lock resource
+     *
+     * @throws LockConflictedException
+     * @throws LockAcquiringException
+     */
+    public function save(Key $key)
+    {
+        $resource = $this->getKeyResource($key);
+
+        $lockNodePath = self::LOCK_NODE.'/'.$resource; // For example: /PHP_LOCK_NODE/MESSAGE_1234
+        // This makes sure the node dies at the end of a session, thus releasing the lock to avoid starvation
+        $flags = Zookeeper::EPHEMERAL;
+
+        try {
+            $token = $this->getToken($key);
+            $this->createNewLock($lockNodePath, $token, array(), $flags);
+        } catch (LockConflictedException $ex) {
+            throw $ex;
+        } catch (Throwable $throwable) {
+            throw new LockAcquiringException($throwable);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockReleasingException
+     */
+    public function delete(Key $key)
+    {
+        $lockNodePath = self::LOCK_NODE.'/'.$this->getKeyResource($key);
+        try {
+            if ($this->exists($key)) {
+                $this->zookeeper->delete($lockNodePath);
+            }
+        } catch (Throwable $throwable) {
+            // For Zookeeper Ephemeral Nodes, the node will be deleted upon session death. But, if we want to unlock
+            // the lock before proceeding further in the session, the client should be aware of this
+            throw new LockReleasingException($throwable);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exists(Key $key): bool
+    {
+        $lockNodePath = self::LOCK_NODE.'/'.$this->getKeyResource($key);
+        if (!$this->zookeeper->exists($lockNodePath)) {
+            return false;
+        }
+
+        return $this->zookeeper->get($lockNodePath) === $this->getToken($key);
```

Comment:
what if the node is removed (by a conurrent request) just before this line? 

IMHO you can remove the `this->zookeeper->exists` and wrap this `this->zookeeper->get` in a try/catch block

Reply:
The lock can only be removed by the original owner and no other concurrent request can delete it unless someone does it via command line or write a script to delete the nodes.